test/state/state.cpp

```diff
@@ -0,0 +1,169 @@
+    return cost;
+}
+
+int64_t compute_tx_intrinsic_cost(evmc_revision rev, const Transaction& tx) noexcept
+{
+    static constexpr auto call_tx_cost = 21000;
+    static constexpr auto create_tx_cost = 53000;
+    const bool is_create = !tx.to.has_value();
+    const auto tx_cost = is_create && rev >= EVMC_HOMESTEAD ? create_tx_cost : call_tx_cost;
+    return tx_cost + compute_tx_data_cost(rev, tx.data) + compute_access_list_cost(tx.access_list);
+}
+
+evmc_message build_message(const Transaction& tx, int64_t execution_gas_limit) noexcept
+{
+    const auto recipient = tx.to.has_value() ? *tx.to : evmc::address{};
+    return {
+        tx.to.has_value() ? EVMC_CALL : EVMC_CREATE,
+        0,
+        0,
+        execution_gas_limit,
+        recipient,
+        tx.sender,
+        tx.data.data(),
+        tx.data.size(),
+        intx::be::store<evmc::uint256be>(tx.value),
+        {},
+        recipient,
+    };
+}
+}  // namespace
+
+std::optional<std::vector<Log>> transition(
+    State& state, const BlockInfo& block, const Transaction& tx, evmc_revision rev, evmc::VM& vm)
+{
+    if (rev < EVMC_LONDON && tx.kind == Transaction::Kind::eip1559)
+        return {};
+
+    if (rev < EVMC_BERLIN && !tx.access_list.empty())
+        return {};
+
+    if (tx.max_gas_price < tx.max_priority_gas_price)
+        return {};  // tip too high
+
+    if (block.gas_limit < tx.gas_limit)
+        return {};
+
+    const auto base_fee = (rev >= EVMC_LONDON) ? block.base_fee : 0;
+    if (tx.max_gas_price < base_fee)
+        return {};
+
+    auto& sender_acc = state.get(tx.sender);
+    if (!sender_acc.code.empty())
+        return {};  // Tx origin must not be a contract (EIP-3607).
+
+    if (sender_acc.nonce == Account::NonceMax)
+        return {};
+
+    const auto tx_max_cost_512 = intx::umul(intx::uint256{tx.gas_limit}, tx.max_gas_price);
+    auto sender_balance_check = sender_acc.balance;
+    if (sender_balance_check < tx_max_cost_512)
+        return {};
+
+    sender_balance_check -= static_cast<intx::uint256>(tx_max_cost_512);
+    if (sender_balance_check < tx.value)
+        return {};
+
+    const auto execution_gas_limit = tx.gas_limit - compute_tx_intrinsic_cost(rev, tx);
+    if (execution_gas_limit < 0)
+        return {};
+
+    assert(tx.max_gas_price >= base_fee);                   // Checked at the front.
+    assert(tx.max_gas_price >= tx.max_priority_gas_price);  // Checked at the front.
+    const auto priority_gas_price =
+        std::min(tx.max_priority_gas_price, tx.max_gas_price - base_fee);
+    const auto effective_gas_price = base_fee + priority_gas_price;
+
+    assert(effective_gas_price <= tx.max_gas_price);
+    const auto tx_max_cost = tx.gas_limit * effective_gas_price;
+
+    sender_acc.balance -= tx_max_cost;  // Modify sender balance after all checks.
+
+    Host host{rev, vm, state, block, tx};
+
+    const auto result = host.call(build_message(tx, execution_gas_limit));
```

Comment:
Where do sender and recipient get their `touched` flag set? inside `call`?

Reply:
In `Host::call()` for `CALL` only: `recipient_acc.touched = true;`.
Sender is not touched.
Are you confused by "touched" vs "warm"?